REGRESSION (252960@main, WPT resync): [ macOS Debug ] imported/w3c/web-platform-tests/html/semantics/document-metadata/the-style-element/style-load-after-mutate.html is a flaky failure

cdumez opened the pull request:
#### a5047de4052aba74f36f2ae51ef33d733aa3b020
<pre>
REGRESSION (252960@main, WPT resync): [ macOS Debug ] imported/w3c/web-platform-tests/html/semantics/document-metadata/the-style-element/style-load-after-mutate.html is a flaky failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=244014">https://bugs.webkit.org/show_bug.cgi?id=244014</a>
rdar://98757197

Reviewed by Ryosuke Niwa.

The test in question starts out with an empty &lt;style&gt; and then adds an @import
to it. However, the URL of the @import points to a missing resource. As a
result, we first schedule the dispatch of a load event, then an error event.
The test makes sure that a load event is fired and times out if none gets fired.

The issue was that HTMLStyleElement was using the same EventSender for the load
and error events. However, EventSender was storing senders in its queue, not
events. The state about whether to send a load or an error event was stored on
the HTMLStyleElement instance instead.

This meant that we could schedule a load event and then an error event right
after one another. The request to send an error event would override the state
on the HTMLStyleElement and we would end up sending 2 error events, instead of
a load event followed by an error event.

To address the issue, I updated EventSender to store the event type in its
queue, in addition to the sender. We would have used two separate EventSenders
like HTMLLinkElement was doing. However, I worry that ordering between load
and error events may be incorrect if we use separate queues.

* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::implicitClose):
* Source/WebCore/dom/EventSender.h:
(WebCore::EventSender::hasPendingEvents const):
(WebCore::Counter&gt;::EventSender):
(WebCore::Counter&gt;::dispatchEventSoon):
(WebCore::Counter&gt;::cancelEvent):
(WebCore::Counter&gt;::dispatchPendingEvents):
(WebCore::EventSender::eventType const): Deleted.
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::didAttachRenderers):
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::linkLoadEventSender):
(WebCore::HTMLLinkElement::~HTMLLinkElement):
(WebCore::HTMLLinkElement::linkLoaded):
(WebCore::HTMLLinkElement::linkLoadingErrored):
(WebCore::HTMLLinkElement::dispatchPendingEvent):
(WebCore::HTMLLinkElement::notifyLoadedSheetAndAllCriticalSubresources):
(WebCore::linkErrorEventSender): Deleted.
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/HTMLStyleElement.cpp:
(WebCore::styleLoadEventSender):
(WebCore::HTMLStyleElement::dispatchPendingEvent):
(WebCore::HTMLStyleElement::notifyLoadedSheetAndAllCriticalSubresources):
* Source/WebCore/html/HTMLStyleElement.h:
* Source/WebCore/html/ImageInputType.cpp:
(WebCore::ImageInputType::attach):
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::loadEventSender):
(WebCore::ImageLoader::ImageLoader):
(WebCore::ImageLoader::~ImageLoader):
(WebCore::ImageLoader::clearImageWithoutConsideringPendingLoadEvent):
(WebCore::ImageLoader::updateFromElement):
(WebCore::ImageLoader::didUpdateCachedImage):
(WebCore::ImageLoader::notifyFinished):
(WebCore::ImageLoader::dispatchPendingEvent):
(WebCore::ImageLoader::dispatchPendingErrorEvent):
(WebCore::beforeLoadEventSender): Deleted.
(WebCore::errorEventSender): Deleted.
(WebCore::ImageLoader::dispatchPendingBeforeLoadEvent): Deleted.
(WebCore::ImageLoader::dispatchPendingBeforeLoadEvents): Deleted.
(WebCore::ImageLoader::dispatchPendingErrorEvents): Deleted.
* Source/WebCore/loader/ImageLoader.h:
(WebCore::ImageLoader::hasPendingBeforeLoadEvent const): Deleted.
* Source/WebCore/svg/animation/SVGSMILElement.cpp:
(WebCore::smilEventSender):
(WebCore::SVGSMILElement::~SVGSMILElement):
(WebCore::SVGSMILElement::progress):
(WebCore::SVGSMILElement::dispatchPendingEvent):
(WebCore::smilBeginEventSender): Deleted.
(WebCore::smilEndEventSender): Deleted.
* Source/WebCore/svg/animation/SVGSMILElement.h:
* Source/WebCore/xml/parser/XMLDocumentParser.cpp:
(WebCore::XMLDocumentParser::append):

Canonical link: <a href="https://commits.webkit.org/256535@main">https://commits.webkit.org/256535@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b72b279654880764d0d875584bd1bac61e42decc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96079 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5328 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29125 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105631 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/165959 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100057 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5446 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34089 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88458 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101448 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101739 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4021 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82685 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/31058 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87779 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/73891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39820 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37495 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/20651 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4522 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/42091 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43981 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39914 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/EventSender.h:
```diff
@@ -37,53 +37,70 @@ class WeakPtrImplWithEventTargetData;
-    Vector<WeakPtr<T, Counter>> m_dispatchSoonList;
-    Vector<WeakPtr<T, Counter>> m_dispatchingList;
+    struct DispatchTask {
+        WeakPtr<T, Counter> sender;
```
The second argument is WeakPtrImpl these days. Rename?

rniwa reviewed (approved):
r=me

cdumez committed:
REGRESSION (252960@main, WPT resync): [ macOS Debug ] imported/w3c/web-platform-tests/html/semantics/document-metadata/the-style-element/style-load-after-mutate.html is a flaky failure
https://bugs.webkit.org/show_bug.cgi?id=244014
rdar://98757197

Reviewed by Ryosuke Niwa.

The test in question starts out with an empty <style> and then adds an @import
to it. However, the URL of the @import points to a missing resource. As a
result, we first schedule the dispatch of a load event, then an error event.
The test makes sure that a load event is fired and times out if none gets fired.

The issue was that HTMLStyleElement was using the same EventSender for the load
and error events. However, EventSender was storing senders in its queue, not
events. The state about whether to send a load or an error event was stored on
the HTMLStyleElement instance instead.

This meant that we could schedule a load event and then an error event right
after one another. The request to send an error event would override the state
on the HTMLStyleElement and we would end up sending 2 error events, instead of
a load event followed by an error event.

To address the issue, I updated EventSender to store the event type in its
queue, in addition to the sender. We would have used two separate EventSenders
like HTMLLinkElement was doing. However, I worry that ordering between load
and error events may be incorrect if we use separate queues.

* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::implicitClose):
* Source/WebCore/dom/EventSender.h:
(WebCore::EventSender::hasPendingEvents const):
(WebCore::Counter>::EventSender):
(WebCore::Counter>::dispatchEventSoon):
(WebCore::Counter>::cancelEvent):
(WebCore::Counter>::dispatchPendingEvents):
(WebCore::EventSender::eventType const): Deleted.
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::didAttachRenderers):
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::linkLoadEventSender):
(WebCore::HTMLLinkElement::~HTMLLinkElement):
(WebCore::HTMLLinkElement::linkLoaded):
(WebCore::HTMLLinkElement::linkLoadingErrored):
(WebCore::HTMLLinkElement::dispatchPendingEvent):
(WebCore::HTMLLinkElement::notifyLoadedSheetAndAllCriticalSubresources):
(WebCore::linkErrorEventSender): Deleted.
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/HTMLStyleElement.cpp:
(WebCore::styleLoadEventSender):
(WebCore::HTMLStyleElement::dispatchPendingEvent):
(WebCore::HTMLStyleElement::notifyLoadedSheetAndAllCriticalSubresources):
* Source/WebCore/html/HTMLStyleElement.h:
* Source/WebCore/html/ImageInputType.cpp:
(WebCore::ImageInputType::attach):
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::loadEventSender):
(WebCore::ImageLoader::ImageLoader):
(WebCore::ImageLoader::~ImageLoader):
(WebCore::ImageLoader::clearImageWithoutConsideringPendingLoadEvent):
(WebCore::ImageLoader::updateFromElement):
(WebCore::ImageLoader::didUpdateCachedImage):
(WebCore::ImageLoader::notifyFinished):
(WebCore::ImageLoader::dispatchPendingEvent):
(WebCore::ImageLoader::dispatchPendingErrorEvent):
(WebCore::beforeLoadEventSender): Deleted.
(WebCore::errorEventSender): Deleted.
(WebCore::ImageLoader::dispatchPendingBeforeLoadEvent): Deleted.
(WebCore::ImageLoader::dispatchPendingBeforeLoadEvents): Deleted.
(WebCore::ImageLoader::dispatchPendingErrorEvents): Deleted.
* Source/WebCore/loader/ImageLoader.h:
(WebCore::ImageLoader::hasPendingBeforeLoadEvent const): Deleted.
* Source/WebCore/svg/animation/SVGSMILElement.cpp:
(WebCore::smilEventSender):
(WebCore::SVGSMILElement::~SVGSMILElement):
(WebCore::SVGSMILElement::progress):
(WebCore::SVGSMILElement::dispatchPendingEvent):
(WebCore::smilBeginEventSender): Deleted.
(WebCore::smilEndEventSender): Deleted.
* Source/WebCore/svg/animation/SVGSMILElement.h:
* Source/WebCore/xml/parser/XMLDocumentParser.cpp:
(WebCore::XMLDocumentParser::append):

Canonical link: https://commits.webkit.org/256535@main
